signer/src/network/libp2p/swarm.rs

```diff
@@ -70,22 +85,66 @@ impl SignerBehavior {
 
         let bootstrap_config = bootstrap::Config::new(local_peer_id)
             .with_initial_delay(config.initial_bootstrap_delay)
-            .add_seed_addresses(config.seed_addresses);
-
+            .add_seed_addresses(&config.seed_addresses);
         let bootstrap = bootstrap::Behavior::new(bootstrap_config);
 
         Ok(Self {
             gossipsub: Self::gossipsub(&keypair)?,
             mdns,
-            kademlia: Self::kademlia(local_peer_id),
+            kademlia,
             ping: Default::default(),
             identify,
             autonat_client,
             autonat_server,
             bootstrap,
+            connection_limits: Self::connection_limits(&config),
         })
     }
 
+    fn connection_limits(config: &SignerSwarmConfig) -> connection_limits::Behaviour {
+        // The number of signers is the number of signers in the set, minus one
+        // for the local signer. This is used to calculate the connection limits.
+        // Note: We use `max(2)` here to ensure that we always have limits
+        // that support at least two signers.
+        let num_signers = config.num_signers.saturating_sub(1).max(2) as u32;
+
+        // Allow `num_signers` * 3 connections to be established at once. This
+        // allows room for one incoming/outgoing connection to each signer, plus
+        // one additional connection per signer as a buffer for protocols such
+        // as autonat.
+        let max_established = num_signers.saturating_mul(3);
+
+        // Allow `num_signers` * 2 _incoming_ connections to be established at
+        // once. This allows room for one incoming connection from each signer,
+        // plus one additional connection per signer as a buffer for protocols
+        // such as autonat.
+        let max_established_incoming = num_signers.saturating_mul(2);
+
+        // Allow for one incoming, one outgoing and one additional connection per
+        // peer.
+        let max_established_per_peer = 3;
```

Comment:
This seems the most "interesting" one. What about bumping it up (eg, 10?) to allow for some slack? For no specific reason, it's just that exact limits seems error prone.

Reply:
Well, 1 extra connection with an already possible usage (autonat) doesn't leave much of a slack. I'm mostly worried about cases where the connection itself is left "established" (eg by some intermediate proxy) with the peer trying to establish a new one being prevent from by this limit. `10` was definitely a made up number, but maybe a 5 (2 normal + 2 new connection buffer + 1 autonat and co) wouldn't be too bad.

Reply:
Mmm but I don't think we should account for that tbh, this doesn't represent a socket (which can be left in some indeterminate state), it's the actual libp2p fully-established connections. A proxy wouldn't be handling the keepalives and so the connection would be dropped after (the now) 10s default..